src/main/java/com/google/cloud/tools/DeployAction.java

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.tools;
+
+import com.google.cloud.tools.ProcessCaller.Tool;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Deploys an app to the server.
+ */
+public class DeployAction extends Action {
+
+  private static Option[] acceptedFlags = {
+      Option.BUCKET,
+      Option.DOCKER_BUILD,
+      Option.FORCE,
+      Option.IMAGE_URL,
+      Option.PROMOTE,
+      Option.SERVER
+  };
+  private AppType appType;
+  private Action stageAction;
```

Comment:
A staged app is used by both deploy and run (dev-appserver).  Rerunning that step for each run/deploy doesn't feel like the most efficient way to handle this. If no changes are detected, we shouldn't have to re-stage.

It's fine if a tool developer wants it to be invisible to the user but we shouldn't make all tool developers follow that flow, and some flexibility in the tool would be helpful in allowing each tool developer to provide an optimized experience. In the case of gradle for instance, I would like stage to be a explicit build step, so it can be skipped when possible.

Another thing is identifying issues, it's more obvious when a failure happens if someone can see which step is failing. Stage and Deploy are separate processes with their own errors, interacting with different systems. I can see an argument for better error messaging here, but the previous point is still relevant to me.